### DISCUSSION TITLE
WPF控件内部结构

### DISCUSSION THREAD

**Discussion by Yangmengbo** [BYJRK/DotNet-Discussions#350] [Q&A]
我要怎么看一个控件的内部结构呢？例如button如果内容为文字的话，它的结构为border=>contentPresent=>TextBlock

**Comment by BYJRK** [comment]
你想在什么时候看？如果是设计时，可以编辑按钮的模板，就可以看到它内部的结构了。如果是运行时，你可以在较新版本的 VS 中找到查看可视化树的功能，或者使用老牌工具 snoop

**Reply by Yangmengbo** [reply]
晚上好，佬，感谢回答。我是昨天看了下b站那个推箱子的视频，里面对ItemsControl进行了很多的修改，但是我右键ItemsControl=》编辑模板=》编辑副本得到的内容很少：    <Window.Resources>
        <Style x:Key="ItemsControlStyle1" TargetType="{x:Type ItemsControl}">
            <Setter Property="Template">
                <Setter.Value>
                    <ControlTemplate TargetType="{x:Type ItemsControl}">
                        <Border Background="{TemplateBinding Background}" BorderBrush="{TemplateBinding BorderBrush}" BorderThickness="{TemplateBinding BorderThickness}" Padding="{TemplateBinding Padding}" SnapsToDevicePixels="true">
                            <ItemsPresenter SnapsToDevicePixels="{TemplateBinding SnapsToDevicePixels}"/>
                        </Border>
                    </ControlTemplate>
                </Setter.Value>
            </Setter>
        </Style>
    </Window.Resources>。就这么点，这似乎不是ItemsControl的全部内容。
另外参考 https://www.codeproject.com/Articles/21495/Understanding-the-Visual-Tree-and-Logical-Tree-in 最后展现出的ListBox的结构我使用snoop查看也没有这么的详细。还有这个https://blog.csdn.net/u012563853/article/details/126178170，它使用的也是snoop，但是我查看treeview也没有这么的详细。vs的可视化树也仅仅是控件的一些摆放的逻辑结构。
